### PR TITLE
Mirror is up only if it has received WAL

### PR DESCRIPTION
**Pull request by asimrp**
Previously, FTS used to treat a mirror as up as soon as the walsender enters `catchup` state.  This patch restricts that criterion.  A mirror is now considered up only if it at least one chunk of WAL has been transferred to it.

An active replication slot is obsoleted when replication lag grows beyond `max_slot_wal_keep_size` limit.  This is a point of no return, the mirror must be destroyed and recreated using `pg_basebackup` (`gprecoverseg -F`).  The walsender process exits when this happens.  However, the walreceiver on mirror is oblivious to this and keeps on attempting new replication connections.

Walsender enters `catchup` state immediately upon receiving a connection request.  The request fails soon after, when an attempt to locate the requested start LSN in locally available WAL fails.  An FTS probe within this window used to erroneously treat the mirror as up and turn syncrep on.  This is definitely not desirable.  The patch causes FTS to treat the mirror as down even if the walsender is found in `catchup` state.

### CODE REVIEW

**Comment by asimrp** [comment]
This PR is a revised version of #12106, based on review feedback.

**Comment by asimrp** [comment]
I still feel that the code is easier to reason about when walsender's state is the only way to tell if a mirror is up or down.  However, without changing the walsender state machine as inherited from PostgreSQL, this is not possible.

**Comment by ashwinstar** [comment]
CI reported unit test failure:
```
                No entries for symbol FindGpdbWalSnd.
1279                ERROR: ../../../../src/test/unit/mock/backend/replication/gp_replication_mock.c:162 - Could not get value to mock function FindGpdbWalSnd
1280                There were no previously returned mock values for this test.
1281[      FAILED ] test_HandleFtsWalRepProbePrimary
1282LOG:  turning off synchronous wal replication due to FTS request
1283                No entries for symbol FindGpdbWalSnd.
1284                ERROR: ../../../../src/test/unit/mock/backend/replication/gp_replication_mock.c:162 - Could not get value to mock function FindGpdbWalSnd
1285                Previously returned mock value was declared at ftsmessagehandler_test.c:106
1286[      FAILED ] test_HandleFtsWalRepSyncRepOff
```

**Comment by ashwinstar** [comment]
> I still feel that the code is easier to reason about when walsender's state is the only way to tell if a mirror is up or down. However, without changing the walsender state machine as inherited from PostgreSQL, this is not possible.

Well yes and no. For this case yes based on replication slot its easier to detect the issue on primary side and reflect the same in state machine. Lets take another example, where nothing is wrong on primary side (hence can't take any action in walsender state machine), just mirror has some problem writing a WAL record (or replaying a WAL record) and keeps crashing and connecting back. How would that situation possibly be handled with walsender state machine. Even with current method possibility exist we may miss these kind of scenarios (if problematic record replay happens after mirror has replied back to primary, but chances of that seem much lower).

The best solution I think would be on mirror side to detect non-correctable errors (without some user / tool action) like corrupted xlog record detection on replay, missing wal files message from primary. Based on these stop auto-retries to primary. Instead once someone takes correctable action trigger the retry. For upstream this would be like restoring WAL from archive for missing WAL and then letting mirror to connect back.

**Comment by soumyadeep2007** [comment]
> Walsender enters catchup state immediately upon receiving a connection request.
Did you mean walreceiver?

**Comment by asimrp** [comment]
No @soumyadeep2007, I meant walsender.  It is walsender where the state machine for streaming replication with a walreceiver is maintained.

**Comment by asimrp** [comment]
@ashwinstar +1 to accomplish the change on the walreceiver side.  However, libpq protocol doesn't currently provide a way to return an indication of an irrecoverable error, so that future connection attempts should not be made.  If a way is carved out, the walreceiver needs to propagate the irrecoverable error to the startup process so that `WaitForWALToBecomeAvailable` no longer attempts to request streaming.  This sounds complex.

I would like to clarify the example you've cited above: 
> Lets take another example, where nothing is wrong on primary side (hence can't take any action in walsender state machine), just mirror has some problem writing a WAL record (or replaying a WAL record) and keeps crashing and connecting back. How would that situation possibly be handled with walsender state machine.

Note that once the mirror disconnects, the state machine on walsender is reset.  Any new connections again starts in catchup state.  Therefore, changing the state machine so that catchup state is attained only after existence of requested startpoint is confirmed, should solve the problem in this particular example.  Am I missing something?

**Comment by ashwinstar** [comment]
> @ashwinstar +1 to accomplish the change on the walreceiver side. However, libpq protocol doesn't currently provide a way to return an indication of an irrecoverable error, so that future connection attempts should not be made. If a way is carved out, the walreceiver needs to propagate the irrecoverable error to the startup process so that `WaitForWALToBecomeAvailable` no longer attempts to request streaming. This sounds complex.

The approach is to just to fully provide context of my mind, I am for sure not proposing to take up that approach currently and what this PR accomplishes seems good approach so far. The comment is more on your discomfort for the solution to be not based on walsender's state machine.

> 
> I would like to clarify the example you've cited above:
> 
> > Lets take another example, where nothing is wrong on primary side (hence can't take any action in walsender state machine), just mirror has some problem writing a WAL record (or replaying a WAL record) and keeps crashing and connecting back. How would that situation possibly be handled with walsender state machine.
> 
> Note that once the mirror disconnects, the state machine on walsender is reset. Any new connections again starts in catchup state. Therefore, changing the state machine so that catchup state is attained only after existence of requested startpoint is confirmed, should solve the problem in this particular example. Am I missing something?

The example I gave is not the one where required WAL is deleted on primary. It's example where nothing is wrong on primary side. Only on mirror side the issue manifests like we had seen where checkpointer process recursively encounters PANIC on fsync request or walreceiver keeps PANICing when writing WAL to disk. All those incidents walsender state machine is not the right tool as it can't detect or act on. It has to be the mirror required to shut-off itself or something based on mirror to convey back to not consider it as up/healthy. It doesn't seem walsender's state machine job.

**Comment by asimrp** [comment]
What if the replication slot is dropped altogether at the time of checkpoint, instead of invalidating its restart_lsn?  What use is such a replication slot anyway?  If it is dropped altogether, subsequent connection attempts from mirror will fail much earlier, at `IDENTIFY SYSTEM` step, because the replication slot name is specified by the mirror in that step.

**Comment by ashwinstar** [comment]
> What if the replication slot is dropped altogether at the time of checkpoint, instead of invalidating its restart_lsn? What use is such a replication slot anyway? If it is dropped altogether, subsequent connection attempts from mirror will fail much earlier, at `IDENTIFY SYSTEM` step, because the replication slot name is specified by the mirror in that step.

Interesting, that feels better for this use-case. As GPDB will always have to perform full recovery and will recreate the slot. Does `gprecoverseg` fail if the slot is not present as it would try to delete and then recreate the slot I think?

Deletion aspect was not discussed at all in upstream thread. Not sure why, maybe something to do with needing more work on delete compared to marking invalid. I don't know. If there is any work-flow where invalid slot can become valid again then that would break with deletion logic added.

In [1] can read a check being discussed at start of replication based on slot's status and fail the connection request and two open items tracked here [2]. Though its only done for logical replication via commit [3]. Would be worth asking why deletion is not good instead of all this work.

1] https://www.postgresql.org/message-id/20200410.115920.1130303270055020682.horikyota.ntt%40gmail.com
2] https://wiki.postgresql.org/wiki/PostgreSQL_13_Open_Items#Open_Issues (Two issues related to slot-invalidation section)
3] https://git.postgresql.org/gitweb/?p=postgresql.git;a=commit;h=d0abe78d84274cc203f3d117b8006dc2164ca31a

**Comment by asimrp** [comment]
>  An FTS probe within this window used to erroneously treat the mirror as up and turn syncrep on.

@paul-guo- @linwen this is the cause of redness seen recently in 6X CI.  The FTS incorrectly determines the mirror to be up, even if the replication slot on primary has been invalidated.  As a result, `gprecoverseg -F` finds no segments needing recovery and subsequent tests fail because they see a mirror intermittently going down.

**Comment by asimrp** [comment]
Thank you @linwen for testing this PR in your dev pipeline.  There is one assertion failure in seg0 primary log:
```
2021-06-11 07:02:23.314162 UTC,"gpadmin",,p15693,th-1399073216,"10.254.0.18","50640",2021-06-11 07:02:23 UTC,0,,,seg0,,,,,"FATAL","XX000","Unexpected internal error (gp_replication.c:476)","FailedAssertion(""!(walsender->is_for_gp_walreceiver)"", File: ""gp_replication.c"", Line: 476)","Process 15693 will wait for gp_debug_linger=120 seconds before termination.
Note that its locks and other resources will not be released until then.",,,,,0,,"gp_replication.c",476,"Stack trace:
1    0x55d8034d3edb postgres errstart + 0x1fb
2    0x55d8034c8e2c postgres ExceptionalCondition + 0x2c
3    0x55d8032e0713 postgres GetMirrorStatus + 0x1e3
4    0x55d803177d4e postgres HandleFtsMessage + 0x24e
5    0x55d803339d0c postgres PostgresMain + 0x21fc
6    0x55d802e7ffa0 postgres <symbol not found> + 0x2e7ffa0
7    0x55d8032a0f83 postgres PostmasterMain + 0x1163
8    0x55d802e9bad3 postgres main + 0x4c3
9    0x7f6da9181bf7 libc.so.6 __libc_start_main + 0xe7
10   0x55d802ea7a0a postgres _start + 0x2a
```

**Comment by asimrp** [comment]
Fixed the assertion failure, it was a mistake on my part.  The latest commit shows how silly that mistake was.

**Comment by ashwinstar** [comment]
> > What if the replication slot is dropped altogether at the time of checkpoint, instead of invalidating its restart_lsn? What use is such a replication slot anyway? If it is dropped altogether, subsequent connection attempts from mirror will fail much earlier, at `IDENTIFY SYSTEM` step, because the replication slot name is specified by the mirror in that step.
> 
> Interesting, that feels better for this use-case. As GPDB will always have to perform full recovery and will recreate the slot. Does `gprecoverseg` fail if the slot is not present as it would try to delete and then recreate the slot I think?
> 
> Deletion aspect was not discussed at all in upstream thread. Not sure why, maybe something to do with needing more work on delete compared to marking invalid. I don't know. If there is any work-flow where invalid slot can become valid again then that would break with deletion logic added.
> 
> In [1] can read a check being discussed at start of replication based on slot's status and fail the connection request and two open items tracked here [2]. Though its only done for logical replication via commit [3]. Would be worth asking why deletion is not good instead of all this work.
> 
> 1] https://www.postgresql.org/message-id/20200410.115920.1130303270055020682.horikyota.ntt%40gmail.com
> 2] https://wiki.postgresql.org/wiki/PostgreSQL_13_Open_Items#Open_Issues (Two issues related to slot-invalidation section)
> 3] https://git.postgresql.org/gitweb/?p=postgresql.git;a=commit;h=d0abe78d84274cc203f3d117b8006dc2164ca31a

Based on that upstream spirit for logical replication if we do it for physical replication in GPDB should help to solve the current problem at hand (this is inline with what you were attempting before)
```diff
diff --git a/src/backend/replication/walsender.c b/src/backend/replication/walsender.c
index ecead8c7539..41e49092a75 100644
--- a/src/backend/replication/walsender.c
+++ b/src/backend/replication/walsender.c
@@ -593,6 +593,13 @@ StartReplication(StartReplicationCmd *cmd)
             ereport(ERROR,
                     (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
                      (errmsg("cannot use a logical replication slot for physical replication"))));
+
+        if (XLogRecPtrIsInvalid(MyReplicationSlot->data.restart_lsn))
+            ereport(ERROR,
+                    (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+                     errmsg("cannot read from logical replication slot \"%s\"",
+                            cmd->slotname),
+                     errdetail("This slot has been invalidated because it exceeded the maximum reserved size.")));
     }
```
I think this might be not generically acceptable in upstream based on when and how physical slot can be created as can pass `RESERVE_WAL` to false. Though in GPDB we always create with `RESERVE_WAL` to true and hence only time `restart_lsn` will be invalid will be due to wal deletion. Though this solution I think can't be applied to gpdb6 as create replication slot doesn't have option `RESERVE_WAL` to immediately start reserving wal. So, for gpdb6 will have to go with what you have in the PR unfortunately.

File: src/backend/replication/gp_replication.c [hunk 1]
```diff
@@ -251,8 +252,12 @@ FTSReplicationStatusUpdateForWalState(const char *app_name, WalSndState state)
 	/* replication_status must exist */
 	Assert(replication_status);
 
-	if (state == WALSNDSTATE_CATCHUP || state == WALSNDSTATE_STREAMING)
+	FtsResponse res;
+	GetMirrorStatus(MyWalSnd, &res);
```
Comment (ashwinstar) [inline src/backend/replication/gp_replication.c:256]: I am trying to understand, why interface for `GetMirrorStatus()` needs to be modified to pass in `walsender`. `FTSReplicationStatusUpdateForWalState()` is only called if `walsnd->is_for_gp_walreceiver`. So, not seeing the need to modify the interface.

File: src/backend/replication/gp_replication.c [hunk 2]
```diff
@@ -251,8 +252,12 @@ FTSReplicationStatusUpdateForWalState(const char *app_name, WalSndState state)
 	/* replication_status must exist */
 	Assert(replication_status);
 
-	if (state == WALSNDSTATE_CATCHUP || state == WALSNDSTATE_STREAMING)
```
Comment (ashwinstar) [inline src/backend/replication/gp_replication.c:254]: Seem `state` argument is no more used.

File: src/backend/replication/gp_replication.c [hunk 3]
```diff
@@ -472,13 +477,15 @@ is_mirror_up(WalSnd *walsender)
...
+	 * A mirror is up only if it has received at least one chunk of WAL.  The
+	 * CATCHUP state is entered as soon as a connection request from mirror is
+	 * received.  The connection may fail soon after if the requested
+	 * startpoint is not found in the WAL files available on primary.
 	 */
-	bool is_communicating_with_mirror = walsender->state == WALSNDSTATE_CATCHUP ||
-		walsender->state == WALSNDSTATE_STREAMING;
+	bool is_communicating_with_mirror =
+		(walsender->state == WALSNDSTATE_CATCHUP &&
+		 !XLogRecPtrIsInvalid(walsender->write)) ||
```
Comment (ashwinstar) [inline src/backend/replication/gp_replication.c:487]: Just trying to think more on this approach, can there be anytime where mirror restarts and at a point where all the WAL generated on primary is present on mirror and no more WAL (or not one chunk of wal) is generated on primary. Will mirror be marked down in that case unnecessarily (sounds extremely rare though)?

File: src/backend/replication/gp_replication.c [hunk 4]
```diff
@@ -528,44 +535,58 @@ is_probe_retry_needed()
 	return false;
 }
 
+/*
+ * Find the walsender serving Greenplum internal replication connection.
+ */
+WalSnd *
+FindGpdbWalSnd(void)
```
Comment (ashwinstar) [inline src/backend/replication/gp_replication.c:542]: Introducing this interface, makes it finding and checking the state within the walsender not atomic which was the case without it. Hence, there is chance this returns walsender and then its state is changed like not used for gpdb before check is performed on it later. Hence, don't like the interface.

Comment (asimrp) [reply src/backend/replication/gp_replication.c]: Good point, the interface is removed, please check the latest version pushed just now.

File: src/backend/replication/gp_replication.c (hunk 3 above)
Comment (asimrp) [reply src/backend/replication/gp_replication.c:490]: This is not a problem because as soon as walsender finds that no more WAL is there to send, it transitions to "streaming" state.  Mirror will be treated as up and in-sync in this case.

File: src/backend/replication/gp_replication.c (hunk 1 above)
Comment (asimrp) [reply src/backend/replication/gp_replication.c]: The reason for adding that interface was the following comment that forced two copies of the same logic to be maintained.
```C
	 * WalSndSetState() resets replica_disconnected_at for
    	 * below states. If modifying below states then be sure
    	 * to update corresponding logic in WalSndSetState()
```
The latest versions achieves that goal without modifying `GetMirrorStatus()` interface.

File: src/backend/fts/ftsmessagehandler.c [hunk 5]
```diff
@@ -309,6 +309,7 @@ HandleFtsWalRepSyncRepOff(void)
 	ereport(LOG,
 			(errmsg("turning off synchronous wal replication due to FTS request")));
 	UnsetSyncStandbysDefined();
+
```
Comment (ashwinstar) [inline src/backend/fts/ftsmessagehandler.c:312]: Lets avoid this unnecessary change

File: src/backend/replication/gp_replication.c [hunk 6]
```diff
@@ -562,7 +572,7 @@ GetMirrorStatus(FtsResponse *response)
 		SpinLockRelease(&walsender->mutex);
 		break;
 	}
-
+
```
Comment (ashwinstar) [inline src/backend/replication/gp_replication.c:575]: avoid this change

File: src/backend/replication/gp_replication.c (hunk 3 above)
Comment (ashwinstar) [reply src/backend/replication/gp_replication.c:490]: Right, missed that aspect, was too much caught in catchup state :-)

File: src/backend/replication/gp_replication.c (hunk 2 above)
Comment (ashwinstar) [reply src/backend/replication/gp_replication.c:254]: This comment remains to be addressed

File: src/test/isolation2/sql/segwalrep/max_slot_wal_keep_size.sql [hunk 7]
```diff
@@ -174,6 +174,32 @@ SELECT role, preferred_role, status FROM gp_segment_configuration WHERE content
...
+-- Wait for the mirror to make the next connection attempt.
+1: SELECT gp_inject_fault('initialize_wal_sender', 'skip',  dbid)
+   FROM gp_segment_configuration WHERE content=0 AND role='p';
+-- Note that we wait until the fault is triggered twice.  Waiting
+-- until the second trigger guarantees that first connection attempt
+-- is fully processed and the status check that follows is accurate.
+1: SELECT gp_wait_until_triggered_fault('initialize_wal_sender', 2, dbid)
+   FROM gp_segment_configuration WHERE content=0 AND role='p';
+
+-- Validate that walsender does not enter catchup state, once the
```
Comment (ashwinstar) [inline src/test/isolation2/sql/segwalrep/max_slot_wal_keep_size.sql:190]: comment needs to be modified as we are not more fiddling with walsender state

File: src/backend/replication/gp_replication.c (hunk 2 above)
Comment (asimrp) [reply src/backend/replication/gp_replication.c:254]: The `state` variable is reinstated by the commit that removed the `FindGpdbWalSnd()` interface.  I've squashed all commits into one, it should be easier to see that now.

**Review by asimrp — COMMENTED** [review]
Apologies for the whitespace diffs.  I've squashed all commits into one and pushed.  Thanks for the patient and diligent review so far!

**Review by ashwinstar — APPROVED** [review]
LGTM. Thank You.

**Comment by asimrp** [comment]
Thank you @wenwangBJ for noticing the unit test failure.  It should be fixed now.  The fix was to initialize the `write` member of the mocked `WalSnd` to `InvalidXLogRecPtr`.

**Comment by asimrp** [comment]
I haven't reverted commit https://github.com/greenplum-db/gpdb/commit/d9afaf44388647180ac4689824363c3d4f2a12d5 because there are still red jobs in 6X_STABLE CI.
